man/io_uring_internals.7

```diff
@@ -0,0 +1,225 @@
+required CPU time in kernel space. The mechanism that
+.I io_uring
+utilizes to achieve this differs depending on the configuration with different
+trade-offs between configurations in respect to e.g. CPU efficiency and latency.
+
+With the default configuration the primary mechanism to provide the kernel space
+CPU time in
+.I io_uring
+is also a syscall: 
+.IR io_uring_enter (2)
+This still differs from requests made via their respective syscall directly,
+such as
+.IR read (2),
+in the sense that it allows for batching in a more flexible way than e.g.
+possible via
+.IR readv (2),
+as different syscalls types can be freely mixed and matched and chains of
+dependent requests, such as a
+.IR send (2)
+followed by a
+.IR recv (2)
+can be submitted with one syscall. Furthermore it is possible to both process
+requests for submissions and process arrived completions within the same
+.IR io_uring_enter (2)
+call. Applications can set the flag
+.I IORING_ENTER_GETEVENTS
+to in addition to processing any pending submissions, process any arrived
+completions and
+optionally wait until a specified amount of completions have arrived before
+returning.
+
+If polled I/O is used all completion related work is performed during the
+.IR io_uring_enter (2)
+call. For interrupt driven I/O, the CPU receiving the hardware interrupt
+schedules the remaining work to be performed including posting the CQE to be
+performed via task work. Any outstanding task work is performed during any
+user-kernel space transition. Per default, the CPU that received the hw
+interrupt will after scheduling the task work interrupt a user space process
+via an inter processor interrupt (IPI), which will cause it to enter the kernel,
+and thus perform the scheduled work. While this ensures a timely delivery of
+the CQE, it is a relatively disruptive and high overhead operation. To avoid
+this applications can configure
+.I io_uring
+via
+.I IORING_SETUP_COOP_TASKRUN
+to elide the IPI. Applications must now ensure that they perform any syscall
+ever so often to be able to observe new completions, but benefit from eliding
+the overheads of the IPIs. Additionally
+.I io_uring
+can be configured to inform an application about the fact that it should now
+perform any syscall to reap new completions by setting
+.IR IORING_SETUP_TASKRUN_FLAG .
+This will result in
+.I io_uring
+setting
+.I IORING_SQ_TASKRUN
+in the SQ flags once the application should do so. This mechanism can be
+restricted further via
+.IR IORING_SETUP_DEFER_TASKRUN ,
+which results in the task work only being executed when
+.IR io_uring_enter (2)
+is called with
+.I IORING_ENTER_GETEVENTS
+set, rather than at any context switch, which gives the application more agency
+about when the work is executed, thus enabling e.g. more opportunities for
+batching.
+
+.PP
+.SH Submission Queue Polling
+.PP
+
+Sq polling introduces a dedicated kernel thread that performs essentially all
```

Comment:
Here, and in other spots, it's important to note that io_uring does NOT utilize any kernel threads. In Linux, a kernel thread is a special kind of thread that is entirely decoupled from any other real process/thread running in userspace. It doesn't have any files, mm, etc associated with it.

What io_uring uses are "io threads", which are exactly like a thread created with eg pthread_create() in the sense that they share any resources that the original task has, and any credentials, namespaces, etc. The only thing that makes them different is that they are created by io_uring, and they never exit to userspace. They sit around and do work, if needed, and then go away when they are no longer needed.

Hence I would probably explain this io thread concept when it's initially encountered in this man page, and then subsequently refer to io threads and remove any mention of kernel threads.

Reply:
Ah I see, my mistake. I thought: thread that never exits to user space i.e. only runs in kernel space == kernel thread. The distinction makes a lot of sense though. I assume this is done for permission management on files (credentials) and zero copy stuff (mm)? I only noticed the difference till now, by io_urings worker threads showing up in traces with tracy or perf, where this did not seem to be the case for other (or what i thought were) other kernel threads.